### ISSUE TITLE
For templated regions, L007 complains that Operators near newlines should be after, not before the newline

### ISSUE THREAD

**Issue by onesuper** [sqlfluff/sqlfluff#3064]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I use sqlfluff to lint my dbt project.

### Expected Behaviour

I assume there should be no error when using a templated region in the Where condition clause

### Observed Behaviour

When a Where condition clause is templated then an L007 error is raised.

### How to reproduce

Using the dbt templator and the included macro lint the following sql:

```SQL
{% macro binary_literal(expression) %}
  X'{{ expression }}'
{% endmacro %}

select
  *
from my_table
where
  a = {{ binary_literal("0000") }}
```

### Dialect

SparkSQL

### Version

```
sqlfluff: 0.12.0
Python 3.9.7
dbt  1.0.4
sqlfluff-templater-dbt : 0.12.0
```

### Configuration

```
[sqlfluff]
dialect = sparksql
templater = dbt

[sqlfluff:templater:dbt]
project_dir = ./

```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
I think I agree with you, but it could be argued that it is right to complain.

Using your templated code, you end up with this SQL:

```sql
select
  *
from my_table
where
  a = 
  X'0000'
```

So the operator (`=`) is before the new line. Which is not what the default setting for L007 is. Hence it flags.

If you don't want this to flag, then shouldn't your macro swallow up the newline with `-%}` syntax to prevent it being included:

```
{% macro binary_literal(expression) -%}
  X'{{ expression }}'
{% endmacro %}
```

Or just be on one line:

```
{% macro binary_literal(expression) -%}X'{{ expression }}'{% endmacro %}
```

Or alternatively disable L007, or configure it to set `operator_new_lines = before`.

I presume you're assuming the, as it's templated code, it doesn't really matter and it should leave it as is. That's not an unreasonably expectation, but think it's slightly complicated, by the fact that the error is with the `=` operator (which is **not** templated code), even though it's basically only being flagged by the newline (which **is** in the templated code).

I think I agree with you it should not flag, but @barrywhart you know the templater MUCH better than me, so any thoughts here?

**Comment by barrywhart** [comment]
This is similar to issues we've had (and fixed) with other rules. Basically, we have to make the rule smarter, to notice it's operating on templated code and do the right thing. This often means mapping back to the corresponding source code to see if that text also violates the rule.

I agree this can be avoided by whitespace stripping, but from a user perspective, I don't think most users want to bother making the **generated** code look good, because they rarely look at it (e.g. only when debugging).

**Comment by onesuper** [comment]
@tunetheweb Thank you very much! I now know what is happening to my templated code.

@barrywhart, I agree with you that the end-users rarely look at it. And it's helpful to let the user know the error is raised from the generated code and be able to ignore the flag. And for a library (macro) developer, generating neat code is necessary.
